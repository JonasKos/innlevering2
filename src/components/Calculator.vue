<template>
  <div class="calc-display">
    <div class="calc">
      <h1>{{ title }}</h1>
      <p class="display">{{ calculation }}</p>
      <div class="calcbuttons">
        <button
          class="calcbtn"
          v-for="variant in variants"
          :style="{ backgroundColor: variant.color }"
          @click="updateCalculation(variant.sign)"
          :key="variant.id"
        >
          {{ variant.sign }}
        </button>
      </div>
    </div>
    <div class="log">
      <h1>Log</h1>
      <ul>
        <li></li>
        <li v-for="log in logs" :key="log.id">{{ log }}</li>
      </ul>
    </div>
  </div>
</template>

<script>
export default {
  name: "HelloWorld",
  props: {
    msg: String,
  },
  data() {
    return {
      title: "Calculator",
      calculation: "",
      calculationLog: "",
      answer: "0",
      logs: [],
      variants: [
        { id: 1, sign: "C" },
        { id: 2, sign: "ANS" },
        { id: 3, sign: "DEL" },
        { id: 4, sign: "+" },
        { id: 5, sign: "1" },
        { id: 6, sign: "2" },
        { id: 7, sign: "3" },
        { id: 8, sign: "-" },
        { id: 9, sign: "4" },
        { id: 10, sign: "5" },
        { id: 11, sign: "6" },
        { id: 12, sign: "*" },
        { id: 13, sign: "7" },
        { id: 14, sign: "8" },
        { id: 15, sign: "9" },
        { id: 16, sign: "/" },
        { id: 17, sign: " " },
        { id: 18, sign: "0" },
        { id: 19, sign: "." },
        { id: 20, sign: "=", color: "orange" },
      ],
    };
  },
  methods: {
    updateCalculation(sign) {
      if (sign === "=") {
        this.calculationLog = this.calculation;
        this.calculation = eval(this.calculation);
        this.logs.push(this.calculationLog + " = " + this.calculation);
        this.answer = this.calculation;
        this.answerTest = this.calculation;
      } else if (sign === "C") {
        this.calculation = "";
      } else if (sign === "DEL") {
        this.calculation = this.calculation.slice(0, -1);
      } else if (sign === "ANS") {
        this.calculation += this.answer;
      } else if (this.calculation == this.answerTest) {
        this.calculation = "";
        this.calculation += sign;
        this.answerTest = "0";
      } else this.calculation += sign;
    },
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style scoped>
@font-face {
  font-family: "digital-clock-font";
  src: url("../assets/digital-7.ttf");
}
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: square;
  padding: 0;
  margin: 0;
}
li {
  min-width: 242px;
  list-style: none;
  padding: 5px;
  margin: 5px;
}
a {
  color: #42b983;
}
.calc-display {
  display: grid;
  justify-content: center;
}
.display {
  font-family: "digital-clock-font";
  font-size: xx-large;
  min-height: 1.25em;
  text-align: right;
  background-color: grey;
  color: white;
  padding: 10px;
  margin: 2px;
}
.calcbuttons {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: auto;
  align-content: center;
  background-color: #ffffff;
}
.calcbtn {
  background-color: grey;
  color: white;
  margin: 2px;
  padding: 10px;
  font-size: 16px;
  text-align: center;
  border-radius: 8px;
}
.log {
  margin-top: 10px;
  border: 1px solid black;
  padding-top: 25px;
  min-height: 300px;
  max-height: 300px;
  scroll-behavior: smooth;
  overflow: auto;
  background-color: rgb(255, 255, 255);
  color: rgb(0, 0, 0);
  font-family: sans-serif;
  padding: 10px;
}
</style>
