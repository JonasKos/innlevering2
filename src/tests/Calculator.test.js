import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import Calculator from "src/components/Calculator.vue";

describe("Calculator.vue", () => {
  it("display blank on mount", () => {
    const wrapper = mount(Calculator);
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("");
  });

  it("display 1 when 1 is clicked", async () => {
    const wrapper = mount(Calculator);
    const button = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    await button.trigger('click');
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("1");
  });

  it("display 12 when 1 and 2 are clicked", async () => {
    const wrapper = mount(Calculator);
    const button1 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    const button2 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "2")[0];
    await button1.trigger('click');
    await button2.trigger('click');
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("12");
  });

  it("display 3 when 1, +, 2, and = buttons are pushed", async () => {
    const wrapper = mount(Calculator);
    const button1 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    const button2 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "2")[0];
    const buttonPlus = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "+")[0];
    const buttonEqual = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "=")[0];
    await button1.trigger('click');
    await buttonPlus.trigger('click');
    await button2.trigger('click');
    await buttonEqual.trigger('click');
    
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("3");
  });

  it("display is cleared when C is clicked", async () => {
    const wrapper = mount(Calculator);
    const button1 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    const button2 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "2")[0];
    const buttonClear = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "C")[0];
    await button1.trigger('click');
    await button2.trigger('click');
    await buttonClear.trigger('click');
    
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("");
  });

  it("ANS button displays last answer", async () => {
    const wrapper = mount(Calculator);
    const button1 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    const button2 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "2")[0];
    const buttonPlus = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "+")[0];
    const buttonEqual = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "=")[0];
    const buttonClear = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "C")[0];
    const buttonAns = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "ANS")[0];
    await button1.trigger('click');
    await buttonPlus.trigger('click');
    await button2.trigger('click');
    await buttonEqual.trigger('click');
    await buttonClear.trigger('click');
    await buttonAns.trigger('click');
    
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("3");
  });

  it("last input is cleared from display when DEL is clicked", async () => {
    const wrapper = mount(Calculator);
    const button1 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "1")[0];
    const button2 = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "2")[0];
    const buttonDel = wrapper.findAll('[class="calcbtn"]').filter((w) => w.text() === "DEL")[0];
    await button1.trigger('click');
    await button2.trigger('click');
    await buttonDel.trigger('click');
    
    const display = wrapper.find('[class="display"]');
    expect(display.text()).toBe("1");
  });
});