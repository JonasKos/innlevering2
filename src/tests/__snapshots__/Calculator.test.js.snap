// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Calculator.vue > renders a message 1`] = `
"<div data-v-c735fc0c=\\"\\" class=\\"calc-display\\">
  <div data-v-c735fc0c=\\"\\" class=\\"calc\\">
    <h1 data-v-c735fc0c=\\"\\">Calculator</h1>
    <p data-v-c735fc0c=\\"\\" class=\\"display\\"></p>
    <div data-v-c735fc0c=\\"\\" class=\\"calcbuttons\\"><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">C</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">ANS</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">DEL</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">+</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">1</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">2</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">3</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">-</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">4</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">5</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">6</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">*</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">7</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">8</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">9</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">/</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\"> </button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">0</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\">.</button><button data-v-c735fc0c=\\"\\" class=\\"calcbtn\\" style=\\"background-color: orange;\\">=</button></div>
  </div>
  <div data-v-c735fc0c=\\"\\" class=\\"log\\">
    <h1 data-v-c735fc0c=\\"\\">Log</h1>
    <ul data-v-c735fc0c=\\"\\">
      <li data-v-c735fc0c=\\"\\"></li>
    </ul>
  </div>
</div>"
`;
